homeassistant/components/solarlog/sensor.py

```diff
@@ -205,40 +216,54 @@ async def async_setup_entry(
     async_add_entities: AddEntitiesCallback,
 ) -> None:
     """Add solarlog entry."""
-    coordinator = entry.runtime_data
-    async_add_entities(
-        SolarlogSensor(coordinator, description) for description in SENSOR_TYPES
-    )
+    coordinator: SolarLogCoordinator = entry.runtime_data
+
+    # https://github.com/python/mypy/issues/14294
 
+    entities: list[CoordinatorEntity] = [
+        SolarLogCoordinatorSensor(coordinator, sensor)  # type: ignore[misc]  # noqa: PGH003
+        for sensor in SENSOR_TYPES
+        if sensor.key in coordinator.data
+    ]
 
-class SolarlogSensor(CoordinatorEntity[SolarlogData], SensorEntity):
-    """Representation of a Sensor."""
+    device_data = coordinator.data.get("devices", {})
 
-    _attr_has_entity_name = True
+    if device_data != {}:
+        for did in device_data:
+            device_id = int(did)
+            if coordinator.solarlog.device_enabled(device_id):
+                entities.extend(
+                    SolarLogInverterSensor(coordinator, sensor, device_id)  # type: ignore[misc]  # noqa: PGH003
+                    for sensor in SENSOR_TYPES
```

Comment:
Does every inverter have the same data as the controller?

Reply:
No, inverter provide less data points (at the moment solarlog-cli has only implemented yearly consumption(i.e. energy) and power). Some additional data points may follow.

Reply:
Should we duplicate the list to make that difference? Because currently both could contain any key and it would create an entity

Reply:
According to my understanding L238 makes sure, that only those sensors are created for the specific device, that are also available for this device. Therefore a combined list is leaner and a duplication - well, a duplication... But happy to double the code if you prefer.

Reply:
And there are many ways to do this. Another way could be to add a list of devices that the entity supports to the entity description. (the downside here is that you can't change icon, name or disabled by default per entity per device) or create a separate list for the other device (causing code duplication, but gives more freedom).

If there is a lot of overlap, I would pick the first one, but if it's only 1-2-3 entities that are shared, we can just duplicate it

Reply:
The overlap is not yet huge, but it presumably will increase. I will therefore go for option 1.